inference-engine/src/inference_engine/ie_cache_guard.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "ie_cache_guard.hpp"
+#include "ie_common.h"
+
+namespace InferenceEngine {
+
+CacheGuardEntry::CacheGuardEntry(CacheGuard& cacheGuard, const std::string& hash,
+                                 std::shared_ptr<std::mutex> m, std::atomic_int& refCount):
+        m_cacheGuard(cacheGuard), m_hash(hash), m_mutex(m), m_refCount(refCount) {
+    // Don't lock mutex right here for exception-safe considerations
+    m_refCount++;
+}
+
+CacheGuardEntry::~CacheGuardEntry() {
+    m_refCount--;
+    m_mutex->unlock();
+    m_cacheGuard.checkForRemove(m_hash);
+}
+
+void CacheGuardEntry::performLock() {
+    m_mutex->lock();
+}
+
+//////////////////////////////////////////////////////
+
+CacheGuard::~CacheGuard() {
+    IE_ASSERT(m_table.size() == 0);
+}
+
+std::unique_ptr<CacheGuardEntry> CacheGuard::getHashLock(const std::string& hash) {
+    std::unique_lock<std::mutex> lock(m_tableMutex);
+    if (!m_table.count(hash)) {
+        m_table.insert({hash, Item{}});
+    }
+    // Now entry exists, need to increment refCount and acquire lock
+    auto& data = m_table[hash];
```

Comment:
If I remember right, the operator [] automatically creates elements with default constructor if it doesn't exist.

Reply:
Yes, good catch, initial version had more of inside 'if', now it can be removed